### PR TITLE
3095 grade release process

### LINKED ISSUE #3095: Update grading form submission process
We need to implement the final portion of the changes to the grading form which will help instructors understand when students will see grades. This will need to happen in phases as 'Release necessary' will be removed as an option on assignment edit form and Grade Status select menu will be replaced with yes/no switch to release grade to student.

grading-form-buttons

[Trello Card](https://trello.com/c/rkxNlEsW/65-change-the-can-the-student-see-this-grade-descriptor-next-to-the-grading-status-dropdown-so-its-meaning-is-clear)

### PR DESCRIPTION
**Pull request by acousticrobot**
### Status
**READY FOR REVIEW**

### Description

This PR is the logic implementation step of the new grade status refactor. It Simplifies the grading release process on the grade form and assignment show page.  Grades can be marked as "Complete", and when they are released to students, they are set to "Student Visible". These states are handled through boolean fields, `grade.status` becomes obsolete. It uses the same pattern for grading Challenges. It also removes the option to require release on assignments; all assignment are treated with the same logic.

### Deploy Notes

Must follow all migrations and rake tasks in:

#3327 (merged and tasks are run)
#3376 (merged and tasks are run)
#3393 (merged and tasks are run)

### Migrations

NO, all migrations should be added first as stated above.

### Steps to Test or Reproduce

Grade assignments and try setting them to either not complete, complete, or visible to students.

Review in progress grades and use the check boxes to release in progress grades on:
  *  grade status page (`/grading_status`)
  * individual assignment show page
  * group assignment show page

Test the same with group grades.
Test student released grades (release should be managed automatically).
Test the same with grading challenges.

### Impacted Areas in Application

* Grading and releasing grades for assignments
* Grading and releasing grades for challenges
* Importing grades

======================
Closes #3095

### CODE REVIEW

**Comment by mariehooper** [comment]
@acousticrobot - This looks great! I just tweaked the styling a bit, but no accessibility concerns to address either! Really happy with the improved clarity for the grader here! 🎉

File: app/controllers/info_controller.rb [hunk 1]
```diff
@@ -32,14 +32,12 @@ def earned_badges
...
 
-  # Displaying all ungraded, graded but unreleased, and in progress assignment
-  # submissions in the system
+  # Displaying ungraded submissions, and existing grades by status
   def grading_status
     grades = current_course.grades.instructor_modified
     submissions = current_course.submissions.submitted.includes(:assignment, :grade, :student, :group, :submission_files)
     @ungraded_submissions_by_assignment = submissions.ungraded.group_by(&:assignment)
     @resubmissions_by_assignment = submissions.resubmitted.group_by(&:assignment)
-    @unreleased_grades_by_assignment = grades.not_released.group_by(&:assignment)
```
Comment (chcholman) [inline app/controllers/info_controller.rb]: I believe we should still have a section of grades that are not yet student visible?

File: app/views/challenge_grades/_form.html.haml [hunk 2]
```diff
@@ -10,12 +10,9 @@
       .form-hint out of #{points @challenge.full_points} points
 
     .form-item
-      - if @challenge.release_necessary?
-        = f.label :status
-        = f.select :status, ChallengeGrade::STATUSES
-        .form-hint= "Can the #{term_for :team} see this grade?"
-      - else
-        = f.input :status, as: :hidden, :input_html => { :value => "Graded" }
+      = f.label :student_visible
```
Comment (chcholman) [inline app/views/challenge_grades/_form.html.haml]: This should be styled like the switch on the grading form, moved to the bottom/right side of the page, and the label should say "Visible to Students?"

**Comment by chcholman** [comment]
@acousticrobot This is looking really good - I left a couple of notes about tweaks to the grading status page and the challenge grade form. I'm going to try to find some time to check a few linguistic details with Evan, and then will be aiming to merge this together on Tuesday!

**Comment by chcholman** [comment]
@acousticrobot A couple of details after talking to Evan:
- Let's rename the column heading from "Complete" to "Graded"
- Let's update the grading form description to match, saying "Mark as Graded?" and the label should say: "Grade will be marked as Graded" vs. "Grade will be marked as In Progress" (I know that's a term we're using in the interface and not one we're storing)

**Pull request merged**